Google Analytics: Add Support for GA4 Tags with GA4 JS

krymson24 opened the pull request:
Helps with 305-gh-dotcom-manage

## Changes proposed in this Pull Request:
Currently, adding a GA4 from the Google Analytics Plugin, does not result in tracking being sent through to the Google Analytics page.

This is because the incorrect Google Analytics script is being used. GA4 uses a new script, documented here, which is backwards compatible https://developers.google.com/analytics/devguides/collection/upgrade/analyticsjs

These changes identify whether the tag is a GA4 tag, and if so, uses the GA4 JS script. Otherwise, we'll continue to use the UA JS script.

That way, tracking that was used by previous users continues to work, and tracking used my new users will work as well.

GA4 Script in source index
<img width="702" alt="Screen Shot 2020-11-12 at 7 05 14 PM" src="https://user-images.githubusercontent.com/66652282/99012158-b8d14800-251b-11eb-9198-d8ed69c8f9a9.png">

Google Analytics Metrics
<img width="779" alt="Screen Shot 2020-11-12 at 7 05 06 PM" src="https://user-images.githubusercontent.com/66652282/99012189-ca1a5480-251b-11eb-9d6c-1287bbc72766.png">

## Jetpack product discussion
N/A.

## Does this pull request change what data or activity we track or use?
No.

## Testing instructions:
For WooCommerce sites:
1. Spin up a JN website
2. Set up Jetpack
3. Choose a Jetpack plan (Daily Security)
4. Set up a new Google Analytics property with a GA4 tag for your Jetpack site
5. In `https://wordpress.com/marketing/traffic/{site_id}`,  enter your GA4 tag in the Google Analytics Tracking ID
6. Go to your Jetpack site, if you inspect the source index, you should be able to find your GA4 tag, and the GA4 script documented here: https://developers.google.com/analytics/devguides/collection/upgrade/analyticsjs
7. Go to your Google Analytics page, you should see at least one user viewing your site in realtime.
8. All functionality with the UA tags should be preserved

### Testing adding custom gtag events via filter

This implementation [enables addition of custom `gtag` events](https://developers.google.com/analytics/devguides/collection/gtagjs/events) via the filter `jetpack_gtag_universal_commands`. To test this.

* Create an `mu-plugin` on your test site - call it whatever you want. Keep it simple.
* Add the following code to your `mu-plugin` - this hooks into the filter and returns an assoc array of events for inclusion in the `<script>` tag output as custom `gtag()` events:

```php
function add_jetpack_gtag_universal_commands() {

	return array(
		array(
			'event',
			'jetpack_testing_event',
			array(
				'event_category' => 'somecat',
				'event_label'    => 'somelabel',
				'value'          => 'someval',
			),
		),
		array(
			'event',
			'another_jetpack_testing_event',
			array(
				'event_category' => 'foo',
				'event_label'    => 'bar',
				'value'          => 'baz',
			),
		),
	);
}

add_filter( 'jetpack_gtag_universal_commands', 'add_jetpack_gtag_universal_commands' );
```

* Ensure your `mu-plugin` is definitely working on your site - add some debugging output and then load the front page of your site and `View Source` (remember it only loads GA code on the front of your site not in WPAdmin).
* `View Source` and search for `dataLayer` using your browser's "Search" utility.
* You should see something akin to:

```html
<!-- Jetpack Google Analytics -->
<script async src='https://www.googletagmanager.com/gtag/js?id=G-XXXXXXX'></script>
<script>
    window.dataLayer = window.dataLayer || [];
    function gtag() { dataLayer.push( arguments ); }
    gtag( 'js', new Date() );
    gtag( 'config', 'G-XXXXXXX' );
    gtag( "event", "jetpack_testing_event", {"event_category":"somecat","event_label":"somelabel","value":"someval"} );
    gtag( "event", "another_jetpack_testing_event", {"event_category":"foo","event_label":"bar","value":"baz"} );
</script>
<!-- End Jetpack Google Analytics -->
```

* Go to your Google Analytics Dashboard for the property that corresponds with the `G-XXXXX` Measurement ID (aka "Tracking ID") that you used above.
* Go to `All Events` and then create x2 events by clicking on the `Create Event` button twice.
* Name your events as per the config above - namely: `jetpack_testing_event` and `another_jetpack_testing_event`.
* Make sure you save both events.
* Back on your site, go around clicking on a few pages. Do this for about 1 minute. Do it also in Incognito mode if possible.
* Back in GA Dashboard go to `Real-time` and look for the box that reads `Event count by Event name`. 
* Wait for a bit and you should see your events show up.
* One day they will probably also show up in the `All Events` section as well but that takes longer!
* For bonus points also create some more custom events as per https://developers.google.com/analytics/devguides/collection/gtagjs/events so that we can be sure we've covered all the formats required by Google.

## Proposed changelog entry for your changes:
- Support Google Analytics 4 tags with GA4 JS script

krymson24 committed:
[not verified] fix(google-analytics) introduce ga4 js for ga4 tags

kwight commented:
> in /wp-admin, go to Settings --> Google Analytics --> Plugin Settings and enter your GA4 tag in the GA Tracking ID input box, and select on the Global Site Tag Tracking Method.

I didn't find this in `/wp-admin` – I got sent to `https://wordpress.com/marketing/traffic/:site?site=inherent-spider.jurassic.ninja#analytics`, and didn't have a `Global Site Tag Tracking Method` setting?..

<img width="1089" alt="Screen Shot 2020-11-13 at 12 19 37 PM" src="https://user-images.githubusercontent.com/349751/99117619-db5e7200-25aa-11eb-8e34-4085de968213.png">

Hm, I didn't get this to work; I confirmed I was on the right branch, but I don't see why that `substr` check would fail though. It may still be a PEBCAK thing.

jeherve commented on modules/google-analytics/classes/wp-google-analytics-universal.php:
```diff
@@ -65,23 +65,39 @@ public function wp_head() {
+		} else {
+			// phpcs:disable WordPress.WP.EnqueuedResources.NonEnqueuedScript
+			$async_code = "
+				<!-- Jetpack Google Analytics -->
```
Maybe we could have those comments be there for both implementations?

obenland replied:
b79ec5525aaa09babc1a4ac48ffc954a4d2e3c33

jeherve commented on modules/google-analytics/classes/wp-google-analytics-universal.php:
```diff
@@ -65,23 +65,39 @@ public function wp_head() {
+				<!-- End Jetpack Google Analytics -->
+			";
+			// phpcs:enable
+			$async_code                = str_replace( '%tracking_id%', $tracking_code, $async_code );
```
Maybe we could do this only once, for both implementations?

mmtr reviewed (commented):
> install the Google Analytics plugin

> go to Settings --> Google Analytics --> Plugin Settings

Why do we need the Google Analytics plugin for this? Is not this fix intended for addressing our built-in Google Analytics support?

If so, we shouldn't install any plugin. Instead, we should configure the Google Analytics settings through Jetpack > Settings > Traffic > Google Analytics:

<img width="1277" alt="Screen Shot 2020-11-16 at 12 01 19" src="https://user-images.githubusercontent.com/1233880/99245543-068fce00-2804-11eb-9be7-925b66593258.png">

The link there redirects to Calypso as noted by @kwight above, where we can enter the measurement ID provided by GA4:

<img width="973" alt="Screen Shot 2020-11-16 at 12 10 52" src="https://user-images.githubusercontent.com/1233880/99246072-fcba9a80-2804-11eb-85b1-79af4b6cd39b.png">

Once I do that, I see my JN site keeps using the old UA approach:

<img width="1165" alt="Screen Shot 2020-11-16 at 12 15 11" src="https://user-images.githubusercontent.com/1233880/99246436-95e9b100-2805-11eb-9dbe-8560f2737ab5.png">

I made sure my JN site uses this branch:

<img width="359" alt="Screen Shot 2020-11-16 at 12 29 59" src="https://user-images.githubusercontent.com/1233880/99247635-853a3a80-2807-11eb-8ee9-8a833de1f627.png">

krymson24 commented:
> > in /wp-admin, go to Settings --> Google Analytics --> Plugin Settings and enter your GA4 tag in the GA Tracking ID input box, and select on the Global Site Tag Tracking Method.
> 
> I didn't find this in `/wp-admin` – I got sent to `https://wordpress.com/marketing/traffic/:site?site=inherent-spider.jurassic.ninja#analytics`, and didn't have a `Global Site Tag Tracking Method` setting?..
> 
> <img alt="Screen Shot 2020-11-13 at 12 19 37 PM" width="1089" src="https://user-images.githubusercontent.com/349751/99117619-db5e7200-25aa-11eb-8e34-4085de968213.png">
> 
> Hm, I didn't get this to work; I confirmed I was on the right branch, but I don't see why that `substr` check would fail though. It may still be a PEBCAK thing.

I had to install a plugin to get this working, but what @mmtr is suggesting might make sense.

krymson24 commented:
> > install the Google Analytics plugin
> 
> > go to Settings --> Google Analytics --> Plugin Settings
> 
> Why do we need the Google Analytics plugin for this? Is not this fix intended for addressing our built-in Google Analytics support?
> 
> If so, we shouldn't install any plugin. Instead, we should configure the Google Analytics settings through Jetpack > Settings > Traffic > Google Analytics:
> 
> <img alt="Screen Shot 2020-11-16 at 12 01 19" width="1277" src="https://user-images.githubusercontent.com/1233880/99245543-068fce00-2804-11eb-9be7-925b66593258.png">
> 
> The link there redirects to Calypso as noted by @kwight above, where we can enter the measurement ID provided by GA4:
> 
> <img alt="Screen Shot 2020-11-16 at 12 10 52" width="973" src="https://user-images.githubusercontent.com/1233880/99246072-fcba9a80-2804-11eb-85b1-79af4b6cd39b.png">
> 
> Once I do that, I see my JN site keeps using the old UA approach:
> 
> <img alt="Screen Shot 2020-11-16 at 12 15 11" width="1165" src="https://user-images.githubusercontent.com/1233880/99246436-95e9b100-2805-11eb-9dbe-8560f2737ab5.png">
> 
> I made sure my JN site uses this branch:
> 
> <img alt="Screen Shot 2020-11-16 at 12 29 59" width="359" src="https://user-images.githubusercontent.com/1233880/99247635-853a3a80-2807-11eb-8ee9-8a833de1f627.png">

This seems like a reasonable question. I did not know that existed and I followed the context here: 305-gh-dotcom-manage. Dropping this to prioritize https://serenityp2.wordpress.com/2020/11/16/updated-sprint-goals-november-16-27-2020/

kwight commented:
> I had to install a plugin to get this working... I did not know that existed and I followed the context here: 305-gh-dotcom-manage.

Jetpack code is written as modules – by `It appears that Jetpack will need a separate update to the google analytics module`, he means [this code](https://github.com/Automattic/jetpack/tree/160217749bf5b9722f561060cc8f784c0d55ffac/modules/google-analytics).

krymson24 commented:
> > I had to install a plugin to get this working... I did not know that existed and I followed the context here: [305-gh-dotcom-manage](https://href.li/?305-gh-dotcom-manage).
> 
> Jetpack code is written as modules – by `It appears that Jetpack will need a separate update to the google analytics module`, he means [this code](https://github.com/Automattic/jetpack/tree/160217749bf5b9722f561060cc8f784c0d55ffac/modules/google-analytics).

I see. I linked this comment back to the issue so whoever is picking it up isn't misled: 305-gh-dotcom-manage#issuecomment-729391997

obenland committed:
[not verified] Add gtag to legacy GA class

That one is called when we don't have WooCommerce active

obenland committed:
[not verified] Enable universal commands for gtag

obenland committed:
[not verified] Use correct escaping function

obenland commented:
It looks like there are two classes that are responsible to render GA codes. When WooCommerce is enabled, `wp-google-analytics-uinversal.php` and otherwise `wp-google-analytics-legacy.php`.

I updated `wp-google-analytics-legacy.php` to render gtags, which enables it for my Jetpack test site where I don't have WooCommerce running.

`wp-google-analytics-uinversal.php` contains quite a bit of logic for tracking ecommerce activity, including button clicks etc.
If this is supposed to be a quick fix, then I'm not sure I'd consider updating those to work with gtag as part of the scope of this task.

What do you all think?

krymson24 commented:
> It looks like there are two classes that are responsible to render GA codes. When WooCommerce is enabled, `wp-google-analytics-uinversal.php` and otherwise `wp-google-analytics-legacy.php`.
> 
> I updated `wp-google-analytics-legacy.php` to render gtags, which enables it for my Jetpack test site where I don't have WooCommerce running.
> 
> `wp-google-analytics-uinversal.php` contains quite a bit of logic for tracking ecommerce activity, including button clicks etc.
> If this is supposed to be a quick fix, then I'm not sure I'd consider updating those to work with gtag as part of the scope of this task.
> 
> What do you all think?

I think if you've fixed the non-ecommerce side, it's fine. GA4 is a completely different API

kwight commented:
>  I'm not sure I'd consider updating those to work with gtag as part of the scope of this task.

Sounds good to me, one thing at a time (this seems the urgent part) 👍

krymson24 committed:
[not verified] fix(google-analytics) revert changes to ga universal

krymson24 commented:
> It looks like there are two classes that are responsible to render GA codes. When WooCommerce is enabled, `wp-google-analytics-uinversal.php` and otherwise `wp-google-analytics-legacy.php`.
> 
> I updated `wp-google-analytics-legacy.php` to render gtags, which enables it for my Jetpack test site where I don't have WooCommerce running.
> 
> `wp-google-analytics-uinversal.php` contains quite a bit of logic for tracking ecommerce activity, including button clicks etc.
> If this is supposed to be a quick fix, then I'm not sure I'd consider updating those to work with gtag as part of the scope of this task.
> 
> What do you all think?

I reverted the changes to `wp-google-analytics-universal.php`. I also tested this using a JN site.

<img width="629" alt="Screen Shot 2020-11-18 at 8 44 57 PM" src="https://user-images.githubusercontent.com/66652282/99610292-87ef8800-29df-11eb-8d4f-09b223f37703.png">
<img width="591" alt="Screen Shot 2020-11-18 at 8 48 13 PM" src="https://user-images.githubusercontent.com/66652282/99610294-87ef8800-29df-11eb-9d50-7b9862a3b19f.png">

mmtr reviewed (dismissed):
Nice, it works great now. I'm not able to see myself in the Realtime page though, but likely it's because I have not configured correctly the data stream in GA, because the Google Tag Manager script looks fine:

<img width="782" alt="Screen Shot 2020-11-19 at 11 53 51" src="https://user-images.githubusercontent.com/1233880/99657076-f8d58500-2a5d-11eb-8abd-e9d20dfff036.png">

simison commented on modules/google-analytics/classes/wp-google-analytics-legacy.php:
```diff
@@ -77,6 +77,19 @@ public function insert_code() {
+		if ( 'G-' === substr( $tracking_id, 0, 2 ) ) {
+			$this->render_gtag_code( $tracking_id );
+		} else {
+			$this->render_ga_code( $tracking_id );
```
If the new code is backwards compatible, how come we would need to keep the old JS code around?

obenland replied:
The ga side has some custom tracking for 404 and search pages that seemed to be worth preserving. Would gtag back compat cover those?

anomiex replied:
It looks like https://developers.google.com/analytics/devguides/collection/gtagjs/events would cover the 404 tracking, and https://developers.google.com/analytics/devguides/collection/gtagjs/pages#manual_pageviews the search thing?

krymson24 replied:
@anomiex I did more reading on this, and in the 404 tracking case, for UA tags, we do this 
```
$custom_vars[] = "_gaq.push(['_trackEvent', '404', document.location.href, document.referrer]);";
```

I'm not entirely sure why we consider this an "event", but based on Google's documentation -- https://developers.google.com/analytics/devguides/collection/gtagjs/events it does not seem to fall under the domain of an "event". Events seem to be user actions.

So I dug some more and am registering it as an "exception" https://developers.google.com/analytics/devguides/collection/gtagjs/exceptions

Lastly, we are doing this for UA tags:

```
		} elseif (
			is_search()
			&& isset( $_REQUEST['s'] )
		) {
			// Set track for searches, if it's a search, and we are supposed to.
			$track['data'] = sanitize_text_field( wp_unslash( $_REQUEST['s'] ) ); // Input var okay.
			$track['code'] = 'search';
		}

		if ( ! empty( $track ) ) {
			$track['url'] = $this->_get_url( $track );
			// adjust the code that we output, account for both types of tracking.
			$track['url']  = esc_js( str_replace( '&', '&amp;', $track['url'] ) );
			$custom_vars[] = "_gaq.push(['_trackPageview','{$track['url']}']);";
		} else {
			$custom_vars[] = "_gaq.push(['_trackPageview']);";
		}
```

Ultimately, what's being done is tracking the URL , but this is all happening by default on every page: https://developers.google.com/analytics/devguides/collection/gtagjs/pages

So I made no changes here.

anomiex replied:
Fine with me on the 404 thing, I only know what I spotted in the docs at a quick glance.

For the search thing, it looks like it's a bit more complicated. The "else" branch of the second `if` there is the default page tracking. But the "then" branch is tracking a URL other than the current one (unless it's just doing a bunch of work to generate the default URL?). But as I said, I'm not very familiar with this stuff so I'll leave the final decision to you and others.

anomiex commented on modules/google-analytics/classes/wp-google-analytics-legacy.php:
```diff
@@ -123,11 +136,43 @@ public function insert_code() {
+			window.dataLayer = window.dataLayer || [];
+			function gtag() { dataLayer.push( arguments ); }
+			gtag( 'js', new Date() );
+			gtag( 'config', '<?php echo esc_js( $tracking_id ); ?>' );
```
Is `esc_js` correct here? While it seems unlikely, if `$tracking_id` contains `&`, `"`, `<`, or `>` those will be HTML entity-encoded rather than encoded for JavaScript.

It seems to me that `json_encode()` would work better.

krymson24 replied:
Done

krymson24 replied:
I tried the recommendations and I get this:

<img width="380" alt="Screen Shot 2020-11-19 at 11 45 31 PM" src="https://user-images.githubusercontent.com/66652282/99761594-7b912b00-2ac4-11eb-82da-d0ebebcf731e.png">

Reverting the suggestions work as expected.

anomiex replied:
Sorry, I should have explained better.
```suggestion
			gtag( 'config', <?php echo wp_json_encode( $tracking_id ); ?> );
```
is what I meant. Since json_encode already quotes strings, you don't have to include quotes of your own.

krymson24 replied:
Ah, thanks @anomiex! Didn't notice that.

anomiex commented on modules/google-analytics/classes/wp-google-analytics-legacy.php:
```diff
@@ -123,11 +136,43 @@ public function insert_code() {
+			function gtag() { dataLayer.push( arguments ); }
+			gtag( 'js', new Date() );
+			gtag( 'config', '<?php echo esc_js( $tracking_id ); ?>' );
+			<?php echo esc_js( implode( "\r\n", $universal_commands ) ); ?>
```
This also seems wrong. As far as I can tell `$universal_commands` is supposed to be something like
```php
$universal_commands = array(
    "gtag( 'event', 'foo', { 'some': 'data' } );", 
    "gtag( 'event', 'bar', { 'some': 'data' } );", 
);
```
Applying the code you have here would result in invalid JS:
```js
gtag( \'event\', \'foo\', { \'some\': \'data\' } );\ngtag( \'event\', \'bar\', { \'some\': \'data\' } );
```

Personally, I'd probably make `$universal_commands` be more like
```php
$universal_commands = [
    array( 'event', 'foo', array( 'some' => 'data' ) ), 
    array( 'event', 'bar', array( 'some' => 'data' ) ), 
];
```
then have code here turn that into `gtag()` calls with something like
```php
foreach ( $universal_commands as $command ) {
    echo 'gtag( ' . implode( ', ', array_map( 'json_encode', $command ) ) . " );\n";
}
```

krymson24 replied:
Done

getdave replied:
Also the `apply_filters` was wrapped in an `array` so when you tried to add events you ended up with incorrect JS format for `gtag`.

I updated this and it seems to work. See testing instructions I added to the PR desc. 🤞

anomiex commented:
In addition to the inline comments, it would be nice if this added some tests of the new functionality. Particularly for use of the new filter.

kwight commented:
Thanks for the review @anomiex ! We'd like to leave any unit testing for later followup – this is under a big time crunch, and needs to get on dotcom ASAP so we can move Simple sites over to it before the current GA VIP plugin we use is deleted November 30 (p9gi28-3gX-p2#comment-8667).

krymson24 commented:
@anomiex @obenland @kwight @simison @creativecoder @jeherve 

I'm going to do what I can to make the suggested improvements to this PR, but whatever is left at the end of the day is what we should stick with (others feel free to contribute).

The reasons for this decision is that our next Jetpack code freeze is on November 23, which is Monday per p1HpG7-3mE-p2 which leaves today for development, and tomorrow for review and merging.

My development environment is going to be particular slow, because the Google Analytics code is not being saved with my Ngrok website, so I have to develop blind with JN sites. I'm not sure why the code isn't working with my particular set-up, and have tried to debug through with @kwight with no success. I'm considering it a rabbit hole at the moment that prevents me from being productive. Welcoming help here.

![2020-11-19 17 53 56 2](https://user-images.githubusercontent.com/66652282/99734116-4bc83000-2a90-11eb-944f-ed5ec5cb375f.gif)

All in all, I believe what we have here is enough as an incremental improvement, since users aren't seeing _any_ metrics at all. The rest can be added later on.

kwight commented:
> our next Jetpack code freeze is on November 23, which is Monday per p1HpG7-3mE-p2 which leaves today for development, and tomorrow for review and merging... I believe what we have here is enough as an incremental improvement, since users aren't seeing any metrics at all

Sounds good! 👍

krymson24 committed:
[not verified] fix(google-analytics) try using json encode

krymson24 committed:
[not verified] fix(google-analytics) add 404 tracking

krymson24 committed:
[not verified] fix(google-analytics) add logging

krymson24 committed:
[not verified] fix(google-analytics) add logging

krymson24 committed:
[not verified] fix(google-analytics) add logging

krymson24 committed:
[not verified] fix(google-analytics) add logging

krymson24 committed:
[not verified] fix(google-analytics) add logging

krymson24 committed:
[not verified] fix(google-analytics) add logging

krymson24 committed:
[not verified] fix(google-analytics) remove extra parens

krymson24 committed:
[not verified] fix(google-analytics) another approach

krymson24 committed:
[not verified] fix(google-analytics) another approach

krymson24 commented:
I still have not verified that 

```
			foreach ( $universal_commands as $command ) {
				echo 'gtag( ' . implode( ', ', array_map( 'wp_json_encode', $command ) ) . " );\n";
			}
			echo esc_js( $custom_vars );
```

are being output correctly. Will resume tomorrow.

getdave reviewed (commented):
I just tested this as per instructions using the Jetpack Beta testing Plugin on my sandbox site.

I can confirm

1. Script is visible in source.
2. Script is the latest GA version from here https://developers.google.com/analytics/devguides/collection/upgrade/analyticsjs
3. Tag Assistant extension in Chrome detects the tag correctly.
4. I see my visits in `Real Time` area of Google Analytics dashboard.
5. The script is rendered before the closing `</head>` tag as expected.

All in all this seems to be working 👍 

One question - if we're upgrading the script, is it backwards compatible with existing Tracking IDs (now renamed to Measurement ID)? Ie: will it "just work" for users who have already got GA activated?

<img width="755" alt="Screen Shot 2020-11-20 at 10 11 51" src="https://user-images.githubusercontent.com/444434/99788818-93df6500-2b19-11eb-9933-0701c167e4fb.png">

getdave commented:
> I still have not verified that
> 
> ```
> 			foreach ( $universal_commands as $command ) {
> 				echo 'gtag( ' . implode( ', ', array_map( 'wp_json_encode', $command ) ) . " );\n";
> 			}
> 			echo esc_js( $custom_vars );
> ```
> 
> are being output correctly. Will resume tomorrow.

I'll do this for you now.

getdave committed:
[not verified] PHPCS formatting fixes

getdave committed:
[not verified] Remove unwanted wrapping array around universal commands

Allows hook filter consumer to supply nested assoc. array to marry up with Google events.

getdave commented:
> In addition to the inline comments, it would be nice if this added some tests of the new functionality. Particularly for use of the new filter.

I think this is a key point.

krymson24 commented:
> > I still have not verified that
> > ```
> > 			foreach ( $universal_commands as $command ) {
> > 				echo 'gtag( ' . implode( ', ', array_map( 'wp_json_encode', $command ) ) . " );\n";
> > 			}
> > 			echo esc_js( $custom_vars );
> > ```
> > 
> > 
> > are being output correctly. Will resume tomorrow.
> 
> ~I'll do this for you now.~
> 
> @krymson24 I tested this and made a fix to ensure it works and outputs as required. I have updated the testing instructions in this PR to ensure it's covered. You might want to run through those.

Thanks @getdave ! Is there any documentation you can pass along that describes how to create an `mu-plugin`? I've looked all over and this is the best I've found https://www.wpdevguides.com/guides/administration/how-to-create-a-mu-plugin-must-use-plugin but this was a Google search. I thought that this has to do with sandboxing and adding the plugin to your sandbox, but I don't understand how that gets activated on my JN site.

krymson24 committed:
[not verified] fix(google-analytics) force 404

krymson24 committed:
[not verified] fix(google-analytics) force 404

krymson24 committed:
[not verified] fix(google-analytics) force 404

krymson24 commented:
> I just tested this as per instructions using the Jetpack Beta testing Plugin on my sandbox site.
> 
> I can confirm
> 
> 1. Script is visible in source.
> 2. Script is the latest GA version from here https://developers.google.com/analytics/devguides/collection/upgrade/analyticsjs
> 3. Tag Assistant extension in Chrome detects the tag correctly.
> 4. I see my visits in `Real Time` area of Google Analytics dashboard.
> 5. The script is rendered before the closing `</head>` tag as expected.
> 
> All in all this seems to be working 👍
> 
> One question - if we're upgrading the script, is it backwards compatible with existing Tracking IDs (now renamed to Measurement ID)? Ie: will it "just work" for users who have already got GA activated?
> 
> <img alt="Screen Shot 2020-11-20 at 10 11 51" width="755" src="https://user-images.githubusercontent.com/444434/99788818-93df6500-2b19-11eb-9933-0701c167e4fb.png">

@getdave we're actually supporting both scripts depending on the tag used:

```
		if ( 'G-' === substr( $tracking_id, 0, 2 ) ) {
			$this->render_gtag_code( $tracking_id );
		} else {
			$this->render_ga_code( $tracking_id );
		}
```

krymson24 committed:
[not verified] fix(google-analytics) force 404

krymson24 committed:
[not verified] fix(google-analytics) only create exception for 404

krymson24 commented:
I did some more testing on this and forced the 404 exceptions:

<img width="665" alt="Screen Shot 2020-11-20 at 11 02 10 AM" src="https://user-images.githubusercontent.com/66652282/99821530-2be04d00-2b20-11eb-9799-a4e18f126699.png">
<img width="303" alt="Screen Shot 2020-11-20 at 11 04 04 AM" src="https://user-images.githubusercontent.com/66652282/99821537-2edb3d80-2b20-11eb-8c09-e4867c4cd0d2.png">

krymson24 commented:
> I just tested this as per instructions using the Jetpack Beta testing Plugin on my sandbox site.
> 
> I can confirm
> 
> 1. Script is visible in source.
> 2. Script is the latest GA version from here https://developers.google.com/analytics/devguides/collection/upgrade/analyticsjs
> 3. Tag Assistant extension in Chrome detects the tag correctly.
> 4. I see my visits in `Real Time` area of Google Analytics dashboard.
> 5. The script is rendered before the closing `</head>` tag as expected.
> 
> All in all this seems to be working 👍
> 
> One question - if we're upgrading the script, is it backwards compatible with existing Tracking IDs (now renamed to Measurement ID)? Ie: will it "just work" for users who have already got GA activated?
> 
> <img alt="Screen Shot 2020-11-20 at 10 11 51" width="755" src="https://user-images.githubusercontent.com/444434/99788818-93df6500-2b19-11eb-9933-0701c167e4fb.png">

I've verified that everything else works. For anyone looking at this, if someone could send me documentation as to how to create an `mu-plugin`, I can do the testing on this, or if someone could test this, we'll be good to go!

I'll be on later today to finalize. @jeherve I understand code freeze is on Monday -- can I merge this in on Monday?

krymson24 commented:
> > In addition to the inline comments, it would be nice if this added some tests of the new functionality. Particularly for use of the new filter.
> 
> I think this is a key point.

I'm not familiar with testing filters. Is a resource I can look at or existing patterns you can refer me to?

anomiex commented:
> I've verified that everything else works. For anyone looking at this, if someone could send me documentation as to how to create an `mu-plugin`, I can do the testing on this, or if someone could test this, we'll be good to go!

Mu-plugin docs: https://wordpress.org/support/article/must-use-plugins/

The short version is that you just drop a PHP file in the mu-plugins directory.

> @jeherve I understand code freeze is on Monday -- can I merge this in on Monday?

Yes, that will be ok. No later than Monday though.

> I'm not familiar with testing filters. Is there resource I can look at or existing patterns you can refer me to?

If you write a PHPUnit test, you just add the filter with `add_filter()`, call the appropriate method, and verify the result is as expected. You can look in tests/php/modules/google-analytics/ to see the existing tests for this module. Off the top of my head, a test might look something like
```php
function test_filter_jetpack_gtag_universal_commands() {
    // Hijack the option for Jetpack_Google_Analytics_Options::get_tracking_code().
    add_filter(
        'pre_option_jetpack_wga',
        function( $option ) {
            $option['code'] = 'G-XXXXXXX';
            return $option;
        }
    );

    // Filter being tested
    add_filter(
        'jetpack_gtag_universal_commands',
        function () {
            return array(
                ... stuff ...
            );
        }
    );

    ob_start();
    ( new Jetpack_Google_Analytics_Legacy() )->insert_code();
    $actual = ob_get_clean();

    $this->assertContains(
        '... gtag stuff ...',
        $actual
    );
}
```
When the test class is extending WP_UnitTestCase you don't have to worry about cleaning up filters after the test, because WP_UnitTestCase handles that for you.

I'm not familiar with the E2E tests, unfortunately, so I can't help you with them.

krymson24 commented:
> > I've verified that everything else works. For anyone looking at this, if someone could send me documentation as to how to create an `mu-plugin`, I can do the testing on this, or if someone could test this, we'll be good to go!
> 
> Mu-plugin docs: https://wordpress.org/support/article/must-use-plugins/
> 
> The short version is that you just drop a PHP file in the mu-plugins directory.
> 
> > @jeherve I understand code freeze is on Monday -- can I merge this in on Monday?
> 
> Yes, that will be ok. No later than Monday though.
> 
> > I'm not familiar with testing filters. Is there resource I can look at or existing patterns you can refer me to?
> 
> If you write a PHPUnit test, you just add the filter with `add_filter()`, call the appropriate method, and verify the result is as expected. You can look in tests/php/modules/google-analytics/ to see the existing tests for this module. Off the top of my head, a test might look something like
> 
> ```
> function test_filter_jetpack_gtag_universal_commands() {
>     // Hijack the option for Jetpack_Google_Analytics_Options::get_tracking_code().
>     add_filter(
>         'pre_option_jetpack_wga',
>         function( $option ) {
>             $option['code'] = 'G-XXXXXXX';
>             return $option;
>         }
>     );
> 
>     // Filter being tested
>     add_filter(
>         'jetpack_gtag_universal_commands',
>         function () {
>             return array(
>                 ... stuff ...
>             );
>         }
>     );
> 
>     ob_start();
>     ( new Jetpack_Google_Analytics_Legacy() )->insert_code();
>     $actual = ob_get_clean();
> 
>     $this->assertContains(
>         '... gtag stuff ...',
>         $actual
>     );
> }
> ```
> 
> When the test class is extending WP_UnitTestCase you don't have to worry about cleaning up filters after the test, because WP_UnitTestCase handles that for you.
> 
> I'm not familiar with the E2E tests, unfortunately, so I can't help you with them.

Thanks for the context @anomiex! I'm thinking through this and I think I need to create the file in my sandbox? I'm not entirely sure actually, but I tried that, and it didn't work. Or do I need to put the file in `docker/mu-plugins/` of this PR, so that it builds with my JN site? Or do I need to `ssh` into my JN site? 🤔

krymson24 commented:
> > > I've verified that everything else works. For anyone looking at this, if someone could send me documentation as to how to create an `mu-plugin`, I can do the testing on this, or if someone could test this, we'll be good to go!
> > 
> > 
> > Mu-plugin docs: https://wordpress.org/support/article/must-use-plugins/
> > The short version is that you just drop a PHP file in the mu-plugins directory.
> > > @jeherve I understand code freeze is on Monday -- can I merge this in on Monday?
> > 
> > 
> > Yes, that will be ok. No later than Monday though.
> > > I'm not familiar with testing filters. Is there resource I can look at or existing patterns you can refer me to?
> > 
> > 
> > If you write a PHPUnit test, you just add the filter with `add_filter()`, call the appropriate method, and verify the result is as expected. You can look in tests/php/modules/google-analytics/ to see the existing tests for this module. Off the top of my head, a test might look something like
> > ```
> > function test_filter_jetpack_gtag_universal_commands() {
> >     // Hijack the option for Jetpack_Google_Analytics_Options::get_tracking_code().
> >     add_filter(
> >         'pre_option_jetpack_wga',
> >         function( $option ) {
> >             $option['code'] = 'G-XXXXXXX';
> >             return $option;
> >         }
> >     );
> > 
> >     // Filter being tested
> >     add_filter(
> >         'jetpack_gtag_universal_commands',
> >         function () {
> >             return array(
> >                 ... stuff ...
> >             );
> >         }
> >     );
> > 
> >     ob_start();
> >     ( new Jetpack_Google_Analytics_Legacy() )->insert_code();
> >     $actual = ob_get_clean();
> > 
> >     $this->assertContains(
> >         '... gtag stuff ...',
> >         $actual
> >     );
> > }
> > ```
> > 
> > 
> > When the test class is extending WP_UnitTestCase you don't have to worry about cleaning up filters after the test, because WP_UnitTestCase handles that for you.
> > I'm not familiar with the E2E tests, unfortunately, so I can't help you with them.
> 
> Thanks for the context @anomiex! I'm thinking through this and I think I need to create the file in my sandbox? I'm not entirely sure actually, but I tried that, and it didn't work. And `docker/mu-plugins/` of this PR is ignored by Git. Or do I need to `ssh` into my JN site? 🤔

krymson24 commented:
Tested the universal commands:

<img width="659" alt="Screen Shot 2020-11-20 at 6 40 42 PM" src="https://user-images.githubusercontent.com/66652282/99860059-ea6f9200-2b5f-11eb-9f30-6dcfefa22a6b.png">

krymson24 commented:
I've been trying to get the tests to run locally with no success.

I thought that `yarn docker:phpunit` would trigger these tests, but the container just exits without running any tests.

I dug deeper into this and ran the tests inside my Docker container. They also just stop without running any tests. I commented out all the tests except for the Google Analytics one, and they still don't run any tests. I ran `yarn docker:phpunit:multisite` and that works, but it runs a different set of tests. I then tried running the file individually `phpunit test-class.google-analytics.php` with a stack trace of:

```
Fatal error: Uncaught Error: Call to undefined function require_jetpack_file() in /var/www/html/wp-content/plugins/jetpack/tests/php/modules/google-analytics/test-class.google-analytics.php on line 9

Error: Call to undefined function require_jetpack_file() in /var/www/html/wp-content/plugins/jetpack/tests/php/modules/google-analytics/test-class.google-analytics.php on line 9

Call Stack:
    0.0024     506592   1. {main}() /usr/local/bin/phpunit:0
    0.0700    9002096   2. PHPUnit\TextUI\Command::main() /usr/local/bin/phpunit:620
    0.0700    9002208   3. PHPUnit\TextUI\Command->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:162
    0.0717    9004184   4. PHPUnit\TextUI\TestRunner->getTest() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:183
    0.0723    9004184   5. PHPUnit\TextUI\TestRunner->loadSuiteClass() phar:///usr/local/bin/phpunit/phpunit/Runner/BaseTestRunner.php:104
    0.0723    9004224   6. PHPUnit\Runner\StandardTestSuiteLoader->load() phar:///usr/local/bin/phpunit/phpunit/Runner/BaseTestRunner.php:157
    0.0724    9041144   7. PHPUnit\Util\FileLoader::checkAndLoad() phar:///usr/local/bin/phpunit/phpunit/Runner/StandardTestSuiteLoader.php:39
    0.0730    9041608   8. PHPUnit\Util\FileLoader::load() phar:///usr/local/bin/phpunit/phpunit/Util/FileLoader.php:45
    0.0737    9051760   9. include_once('/var/www/html/wp-content/plugins/jetpack/tests/php/modules/google-analytics/test-class.google-analytics.php') phar:///usr/local/bin/phpunit/phpunit/Util/FileLoader.php:57
```

If I can get some guidance from someone, I'll try to write some tests on Monday, but ultimately, I've tested all of the cases manually and they look good to go. Others are welcome to test as well!

mmtr committed:
[not verified] Add unit tests

mmtr commented:
> Others are welcome to test manually or write tests as well!

I added them in a805ba. You can run them locally with `yarn docker:phpunit --filter=WP_Test_Jetpack_Google_Analytics`.

mmtr commented:
Seems that E2E failures are unrelated to this PR.

jeherve reviewed (changes requested):
I'm seeing the views being tracked as expected. I see the eCommerce events are not being fired, but that seems to be beyond the scope of this PR as far as I understand from the comments above. Are y'all tracking this remaining work somewhere, or should I create an issue for it?

The tests are failing right now, though:
https://travis-ci.com/github/Automattic/jetpack/jobs/445786903

kwight commented:
> Is there any documentation you can pass along that describes how to create an mu-plugin?

@krymson24 There's info [in the docs](https://github.com/Automattic/jetpack/blob/master/docker/README.md) about doing this for your Jetpack Docker setup.

krymson24 commented:
> > Others are welcome to test manually or write tests as well!
> 
> I added them in [a56d932](https://github.com/Automattic/jetpack/commit/a56d932dd3ff6ca28c5bbedb18f426699739d64e). You should be able to run them locally with `yarn docker:phpunit --filter=WP_Test_Jetpack_Google_Analytics` after spinning up your Docker container.

Thanks @mmtr ! That works -- and it looks like the tests pass:

```
...                                                                 3 / 3 (100%)
```

Out of curiosity, I also ran `yarn docker:phpunit` because I still expect it to work and I'm finally seeing results.

I don't know what changed. There are failing tests:

```
1) WP_Test_Jetpack_Sync_Themes::test_update_themes_sync
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'Twenty Sixteen'
+'twentysixteen'

/var/www/html/wp-content/plugins/jetpack/tests/php/sync/test_class.jetpack-sync-themes.php:266

2) WP_Test_Jetpack_REST_API_endpoints::test_change_owner
Failed asserting that 500 matches expected 200.

/var/www/html/wp-content/plugins/jetpack/tests/php/_inc/lib/test_class.rest-api-endpoints.php:150
/var/www/html/wp-content/plugins/jetpack/tests/php/_inc/lib/test_class.rest-api-endpoints.php:999

3) WP_Test_Jetpack_Google_Analytics::test_filter_jetpack_gtag_universal_commands
Failed asserting that '' contains "gtag( "event", "jetpack_testing_event", {"event_category":"somecat","event_label":"somelabel","value":"someval"} )".

/var/www/html/wp-content/plugins/jetpack/tests/php/modules/google-analytics/test-class.google-analytics.php:134
```

And it looks like the Google Analytics tests are failing, contrary to the filter test. I'm going to look further into it and fix it if it requires fixing.

krymson24 commented:
> > Is there any documentation you can pass along that describes how to create an mu-plugin?
> 
> @krymson24 There's info [in the docs](https://github.com/Automattic/jetpack/blob/master/docker/README.md) about doing this for your Jetpack Docker setup.

Thanks @kwight! Though that seems like it would only be helpful if I ran Jetpack locally -- and I wasn't able to. I had to run a JN site, but I eventually figured it out by SSHing into the machine. Alternatively, I think I could have removed that directory from git ignore, and pushed the plugin up, but good thing I didn't have to resort to that.

krymson24 commented:
> I'm seeing the views being tracked as expected. I see the eCommerce events are not being fired, but that seems to be beyond the scope of this PR as far as I understand from the comments above. Are y'all tracking this remaining work somewhere, or should I create an issue for it?
> 
> The tests are failing right now, though:
> https://travis-ci.com/github/Automattic/jetpack/jobs/445786903

@jeherve issue created here! https://github.com/Automattic/jetpack/issues/17878

I'll look into the failing tests.

krymson24 committed:
[not verified] fix(google-analytics-tag) add clarification to README

kwight commented:
All of this works wonderfully for me 🎉  I verified:

* a UA code outputs the original script,
* a new code outputs the new script,
* real-time activity looked as expected, and
* custom event tags registered properly.

The one thing to note was that I wasn't seeing the `another_jetpack_testing_event` at first, but it was because of the hook's data format: that third property has to be an array, not a string (changing `bar` for an array of key-pair values made them show up just fine).

Like @krymson24 and @mmtr , I can also see the tests passing locally, but the Travis test that fails indicates maybe a problem with using output buffers in CI?.. I can't try to re-run the tests because the account shows as disabled – @jeherve is there something to be done to reactivate?

krymson24 commented:
> All of this works wonderfully for me 🎉 I verified:
> 
> * a UA code outputs the original script,
> * a new code outputs the new script,
> * real-time activity looked as expected, and
> * custom event tags registered properly.
> 
> The one thing to note was that I wasn't seeing the `another_jetpack_testing_event` at first, but it was because of the hook's data format: that third property has to be an array, not a string (changing `bar` for an array of key-pair values made them show up just fine).
> 
> Like @krymson24 and @mmtr , I can also see the tests passing locally, but the Travis test that fails indicates maybe a problem with using output buffers in CI?.. I can't try to re-run the tests because the account shows as disabled – @jeherve is there something to be done to reactivate?

@kwight thanks for testing and thanks for your approval!

The error that you're referring to might be a red herring though. I am seeing that when I run all the unit tests, it fails with this error:

```
3) WP_Test_Jetpack_Google_Analytics::test_filter_jetpack_gtag_universal_commands
Failed asserting that '' contains "gtag( "event", "jetpack_testing_event", {"event_category":"somecat","event_label":"somelabel","value":"someval"} )".
```

but when run by itself, there's no problems.

@jeherve we'd like to get this merged in today -- can you provide more context as to how I can debug this test case and perhaps why might it not be working when all the tests run, but it works when run in isolation? I've tried adding logs to the file and test files, and adding a `--verbose` flag to the test cases, with no success. With more information, I can try to debug and fix these test cases.

Otherwise, let's remove it and merge.

kwight commented:
> I am seeing that when I run all the unit tests, it fails with this error

Ah, OK, so it does fail locally 🙃 . Hm, well, if you can figure out why, great (I don't _think_ there are problems with using output buffering, particularly because it passes when filtered – maybe change `( new Jetpack_Google_Analytics_Legacy() )->insert_code();` to $legacy = new Jetpack_Google_Analytics_Legacy(); $legacy->insert_code();`?). If not, leave it for @jeherve to decide how how he wants to proceed for us to get this in.

mmtr commented:
> we'd like to get this merged in today

Just wanted to know that we shouldn't be affected by the code freeze. The Nov 30th deadline is for Simple sites, not Jetpack sites, so it doesn't matter if the fix is released as part of Jetpack 9.2 or Jetpack 9.3 as long as this is merged before Nov 30th. As soon as that happens, we'll be able to sync the code to Simple sites.

jeherve commented:
> Just wanted to know that we shouldn't be affected by the code freeze.

The fix needs to happen in Jetpack as well though, so we'll definitely aim to merge this today.

mmtr committed:
Update test

mmtr commented:
> The one thing to note was that I wasn't seeing the `another_jetpack_testing_event` at first, but it was because of the hook's data format: that third property has to be an array, not a string (changing `bar` for an array of key-pair values made them show up just fine).

Yup, that's aligned with the documentation in https://developers.google.com/analytics/devguides/collection/gtagjs/events, so I updated the testing instructions and the part of the unit test that was covering this scenario.

The unit test keeps failing though, but as @krymson24 noted, it happens only when running them all:
- `yarn docker:phpunit --filter=WP_Test_Jetpack_Google_Analytics`: the `test_filter_jetpack_gtag_universal_commands` test passes.
- `yarn docker:phpunit`: the `test_filter_jetpack_gtag_universal_commands` test fails.

mmtr commented:
Found it, there was a preceding test that was setting an admin screen, causing the `insert_code` function to bail:

https://github.com/Automattic/jetpack/blob/11846dfd09fff49eb804baef4ea86351f13324ad/modules/google-analytics/classes/wp-google-analytics-legacy.php#L67-L69

I fixed it by ensuring the current screen is set to the frontend before trying to insert the GA code.

I initially thought that the constructor of the class was also messing up the execution which is why I mocked the class to disable the original class constructor. Although it was unrelated to the test failures, I kept the mock to make the test less prone to errors.

mmtr committed:
Set frontend screen when running test

jeherve reviewed (approved):
Looks good now. 🚢

krymson24 commented:
Thanks @mmtr and @jeherve !
